File: backend/src/Domain/AiStudyOS.Domain/Quiz/Quiz.cs
using AiStudyOS.Domain.Common;

namespace AiStudyOS.Domain.Quiz;

public class Quiz : AggregateRoot
{
    public Guid UserId { get; private set; }
    public Guid? GoalId { get; private set; }
    public string Title { get; private set; } = null!;
    public string Topic { get; private set; } = null!;
    public Difficulty Difficulty { get; private set; }
    public QuizType QuizType { get; private set; }
    public int QuestionCount { get; private set; }
    public string ModelUsed { get; private set; } = null!;
    public string PromptVersion { get; private set; } = null!;
    public string CorrelationId { get; private set; } = null!;
    public DateTime CreatedAtUtc { get; private set; }

    private Quiz() { }

    public static Quiz Create(
        Guid userId,
        Guid? goalId,
        string title,
        string topic,
        Difficulty difficulty,
        QuizType quizType,
        int questionCount,
        string modelUsed,
        string promptVersion,
        string correlationId,
        DateTime nowUtc) => new()
    {
        UserId = userId,
        GoalId = goalId,
        Title = title,
        Topic = topic,
        Difficulty = difficulty,
        QuizType = quizType,
        QuestionCount = questionCount,
        ModelUsed = modelUsed,
        PromptVersion = promptVersion,
        CorrelationId = correlationId,
        CreatedAtUtc = nowUtc,
    };
}

File: frontend/lib/types/quiz.ts
export type QuizTypeValue = "Standard" | "Adaptive" | "Review";
export type QuestionTypeValue = "MultipleChoice" | "TrueFalse" | "ShortAnswer" | "FillBlank";
export type DifficultyValue = "Easy" | "Medium" | "Hard";
export type AttemptStatusValue = "InProgress" | "Completed";

export interface Question {
  id: string;
  order: number;
  type: QuestionTypeValue;
  topic: string;
  difficulty: DifficultyValue;
  text: string;
  options: string[] | null;
}

export interface QuizDetail {
  id: string;
  title: string;
  topic: string;
  difficulty: DifficultyValue;
  quizType: QuizTypeValue;
  questionCount: number;
  createdAtUtc: string;
  questions: Question[];
}

export interface QuizSummary {
  id: string;
  title: string;
  topic: string;
  difficulty: DifficultyValue;
  quizType: QuizTypeValue;
  questionCount: number;
  createdAtUtc: string;
  latestAttemptScore: number | null;
  latestAttemptAtUtc: string | null;
}

export interface AnswerResult {
  questionId: string;
  questionText: string;
  topic: string;
  userAnswer: string;
  correctAnswer: string;
  isCorrect: boolean;
  explanation: string;
}

export interface QuizAttemptResult {
  attemptId: string;
  quizId: string;
  quizTitle: string;
  score: number;
  correctCount: number;
  totalCount: number;
  completedAtUtc: string;
  answers: AnswerResult[];
  weakTopics: string[];
  recommendedTopics: string[];
  estimatedMasteryDelta: number;
  confidence: number;
}

export interface QuizHistoryItem {
  attemptId: string;
  quizId: string;
  quizTitle: string;
  topic: string;
  score: number | null;
  correctCount: number;
  totalCount: number;
  status: AttemptStatusValue;
  startedAtUtc: string;
  completedAtUtc: string | null;
}

export interface TopicMastery {
  topic: string;
  masteryScore: number;
  attemptsCount: number;
  lastUpdatedUtc: string;
}

export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
  hasMore: boolean;
}

export interface GenerateQuizInput {
  topic: string | null;
  goalId: string | null;
  difficulty: DifficultyValue;
  questionTypes: QuestionTypeValue[];
  questionCount: number;
  quizType: QuizTypeValue;
}

export interface SubmittedAnswer {
  questionId: string;
  answer: string;
}

export type QuizGenerationStreamEvent =
  | { type: "delta"; content: string }
  | { type: "complete"; quiz: QuizDetail }
  | { type: "error"; message: string };
